### PR TITLE
add subscription descriptor parser

### PR DESCRIPTION
**Pull request by MohamedFarouk-HMCTS**

### CODE REVIEW

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @JsonProperty("spec_version")
+    private Double specVersion;
+    @JsonProperty("service")
+    private String service;
+    @JsonProperty("service_component")
+    private String serviceComponent;
+    @JsonProperty("subscriptions")
+    private List<Subscription> subscriptions = null;
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptor.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptor.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import com.fasterxml.jackson.annotation.JsonAnySetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptor.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import com.fasterxml.jackson.annotation.JsonAnySetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        "rest_uri"
+})
+public class Location {
+
+    @JsonProperty("jms_uri")
+    private String jmsUri;
+    @JsonProperty("rest_uri")
+    private String restUri;
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptorDef.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+        "subscription_descriptor"
+})
+public class SubscriptionDescriptorDef {
+
+    @JsonProperty("subscription_descriptor")
+    private SubscriptionDescriptor subscriptionDescriptor;
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/SubscriptionDescriptorDef.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import com.fasterxml.jackson.annotation.JsonAnySetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        "location"
+})
+public class Eventsource {
+
+    @JsonProperty("name")
+    private String name;
+    @JsonProperty("location")
+    private Location location;
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Event.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Event.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Event.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import com.fasterxml.jackson.annotation.JsonAnySetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Event.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Subscription.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+public class Subscription {
+
+    @JsonProperty("name")
+    private String name;
+    @JsonProperty("events")
+    private List<Event> events = null;
+    @JsonProperty("eventsource")
+    private Eventsource eventsource;
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Subscription.java]: comment

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Subscription.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Subscription.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Subscription.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Subscription.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Event.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        "schema_uri"
+})
+public class Event {
+
+    @JsonProperty("name")
+    private String name;
+    @JsonProperty("schema_uri")
+    private String schemaUri;
+    @JsonIgnore
+    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Event.java]: done

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.Map;
+public class Location {
+    private String jmsUri;
+    private String restUri;
+    private Map<String, Object> additionalProperties = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java]: removed

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java [hunk 18]
```diff
@@ -0,0 +1,27 @@
+package uk.gov.justice.domain.subscriptiondescriptor;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Eventsource {
+
+    private String name;
+    private Location location;
+    private Map<String, Object> additionalProperties = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java]: removed

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParser.java [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    public Collection<SubscriptionDescriptor> parse(final Path baseDir, final Collection<Path> paths) {
+        return paths.stream()
+                .map(path -> read(baseDir.resolve(path).toAbsolutePath().toString()))
+                .collect(Collectors.toList());
+    }
+
+    private SubscriptionDescriptor read(final String filePath) {
+        final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
```
Comment (mapingo) [inline subscription-descriptor-parser/src/main/java/uk/gov/justice/maven/generator/io/files/parser/SubscriptionDescriptorFileParser.java]: I think we can add the following...then the setters in the classes can be replaced with constructors
```

                .registerModule(new Jdk8Module())
                .registerModule(new ParameterNamesModule(PROPERTIES))
```

File: subscription-descriptor-parser/pom.xml [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.dataformat</groupId>
+            <artifactId>jackson-dataformat-yaml</artifactId>
+            <version>${jackson-dataformat-yaml-version}</version>
+        </dependency>
```
Comment (mapingo) [inline subscription-descriptor-parser/pom.xml:38]: Add these dependencies then the setters can be removed and be replaced with a constructor instead
```
        <dependency>
            <groupId>com.fasterxml.jackson.datatype</groupId>
            <artifactId>jackson-datatype-jdk8</artifactId>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.module</groupId>
            <artifactId>jackson-module-parameter-names</artifactId>
        </dependency>
```

**Review by mapingo — CHANGES_REQUESTED** [review]

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (mapingo) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Location.java]: this still needs to be removed

File: subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (mapingo) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java]: this still needs to be removed

[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply subscription-descriptor-parser/src/main/java/uk/gov/justice/domain/subscriptiondescriptor/Eventsource.java]: removed

**Review by mapingo — APPROVED** [review]

**Pull request closed without merging**